### ISSUE TITLE
ContextObject.view_loop() has a memory leak

### ISSUE THREAD

**Issue by DaFluffyPotato** [cmbruns/pyopenxr#116]
Testing the [pink_world.py](https://github.com/cmbruns/pyopenxr_examples/blob/main/xr_examples/pink_world.py) example results in a memory leak (RAM not VRAM).

I added pympler to verify the source of the leak:
```py
from pympler import tracker
from OpenGL import GL
import xr

# ContextObject is a high level pythonic class meant to keep simple cases simple.
with xr.ContextObject(
    instance_create_info=xr.InstanceCreateInfo(
        enabled_extension_names=[
            # A graphics extension is mandatory (without a headless extension)
            xr.KHR_OPENGL_ENABLE_EXTENSION_NAME,
        ],
    ),
) as context:
    tr = tracker.SummaryTracker()
    
    for frame_index, frame_state in enumerate(context.frame_loop()):
        for view in context.view_loop(frame_state):
            GL.glClearColor(1, 0.7, 0.7, 1)  # pink
            GL.glClear(GL.GL_COLOR_BUFFER_BIT)
        
        if frame_index % 300 == 0:
            tr.print_diff()

```

Which eventually stabilizes after initialization and continuously yields the following diff:
```
                                       types |   # objects |   total size
============================================ | =========== | ============
                                        dict |        6300 |      1.21 MB
                                  memoryview |        1200 |    215.62 KB
                         xr.typedefs.Rect2Di |        1200 |    150.00 KB
               xr.typedefs.SwapchainSubImage |        1200 |    150.00 KB
                               managedbuffer |        1200 |    150.00 KB
                  xr.typedefs.c_long_Array_2 |        1200 |    150.00 KB
  xr.typedefs.CompositionLayerProjectionView |         600 |     75.00 KB
                       xr.typedefs.Extent2Di |         600 |     75.00 KB
                       xr.typedefs.Offset2Di |         600 |     75.00 KB
```

I've watched the program go from ~100MB to several hundred MB of usage before killing it.

Commenting out the following lines from pink_world.py completely removes the leak:
```py
for view in context.view_loop(frame_state):
    GL.glClearColor(1, 0.7, 0.7, 1)  # pink
    GL.glClear(GL.GL_COLOR_BUFFER_BIT)
```

It seems that the contents of `ContextObject.render_layers` are not getting deallocated even though `ContextObject.frame_loop()` does a `self.render_layers = []` prior to yielding each frame.

I believe all the examples including `hello_xr.py` have leaks.

I might look more into the issue later if I have the time.

**Comment by DaFluffyPotato** [comment]
I looked a bit more. It looks like it's the `CompositionLayerProjectionView`s briefly stored in `ContextObject.render_layers` that don't get deallocated. The other xr.typedef items appear to be children of `CompositionLayerProjectionView`.

Here's some useful ref tracing from pympler on CompositionLayerProjectionView:
```
<class 'xr.typedefs.CompositionLayerProjectionView'>(1987851081552)-+-<class 'xr.typedefs.SwapchainSubImage'>(1987851082576)--<class 'xr.typedefs.Rect2Di'>(1987851081168)--<class 'xr.typedefs.Offset2Di'>(1987851082960)--<class 'dict'>(1987825452224)({'_wrapper': xr.Offset2Di(x=0, y=0)})
                                                                    +-<class 'xr.typedefs.SwapchainSubImage'>(1987851083344)--<class 'xr.typedefs.Rect2Di'>(1987851083472)--<class 'xr.typedefs.Extent2Di'>(1987851083600)--<class 'dict'>(1987825452544)({'_wrapper': xr.Extent2Di(width=2064, height=2272)})
```

I had the dictionary print out its contents. I'm guessing it's related to [this line](https://github.com/cmbruns/pyopenxr/blob/main/src/xr/typedefs.py#L2088) in `typedefs.py` in the `Extent2Di` case.

**Comment by cmbruns** [comment]
Thank you for investigating this. It does look like those .as_numpy() methods might be creating a circular reference that inhibits garbage collection. I don't remember at the moment if I created those methods just to allow python sequence-like access e.g. 'foo[key] = bar' for sequency openxr data types, or if there was some other reason to allow persistent access to the result of `as_numpy()`.

**Comment by DaFluffyPotato** [comment]
The deletions I made in the PR fixed the issue for me. I've never had to do lifetime hacks for anything before, so my fix could be very wrong. lol

**Comment by DaFluffyPotato** [comment]
Closing this since #117 was merged.

**Comment by DaFluffyPotato** [comment]
@cmbruns I just figured out what the original intention behind the lifetime hack likely was while toying with the hello_xr example to debug something else. The deallocation of the Matrix4x4f object was messing up the return value of `.as_numpy()`. It's common to use the object just to get the `.as_numpy()` value while never saving the object long term. This can cause issues depending on how Matrix4x4f is used. The fix didn't cause any issues for me since I substitute GLM for all my linear algebra in my projects.

The lifetime hack fundamentally leaks though, so the correct implementation should probably be to either create a copy of the data that isn't dependent on the parent's lifetime, clearly indicate that the validity of the return value is tied to the lifetime of the owning object, or both (as separate functions). The return value does seem to corrupt my draw calls if I delete the lifetime hack and use the return value from a deallocated Matrix4x4f.

I see that matrix4x4f has been changed since I modified it. It looks like it may be a copy now.
https://github.com/cmbruns/pyopenxr/blob/main/src/xr/utils/matrix4x4f.py
